File: BEmu/ReferenceDataRequest/ElementReferenceArrayChainTickers.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/ElementReferenceArrayChainTickers.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementReferenceArrayChainTickers : ElementReferenceArray
    {
        internal enum OptionalityEnum { put = 'P', call = 'C' }

        internal ElementReferenceArrayChainTickers(string underlier, uint numPoints, string strDtExp, OptionalityEnum optionality) : base("CHAIN_TICKERS", new List<Element>())
        {
            //assume that the underlier takes the form "ABC US EQUITY"
            string ticker = underlier.Substring(0, underlier.IndexOf(' '));

            DateTime dtExp;
            if (string.IsNullOrWhiteSpace(strDtExp))
                dtExp = DateTime.Today.AddMonths(1).AddDays(-DateTime.Today.Day).AddDays(20); //assume the 20th of the month
            else if (strDtExp.Length == 8)
                dtExp = DateTime.ParseExact(strDtExp, "yyyyMMdd", System.Globalization.CultureInfo.InvariantCulture);
            else if (strDtExp.Length == 6)
                dtExp = DateTime.ParseExact(strDtExp + "20", "yyyyMMdd", System.Globalization.CultureInfo.InvariantCulture); //assume the 20th of the month
            else if (strDtExp[strDtExp.Length - 1] == 'F')
            {
                uint numMonths = uint.Parse(strDtExp.Substring(0, strDtExp.Length - 1));
                dtExp = DateTime.Today.AddMonths((int)numMonths).AddDays(-DateTime.Today.Day).AddDays(20); //assume the 20th of the month
            }
            else
                throw new NotImplementedException("BEmu.ReferenceDataRequest.ElementReferenceArrayChainTickers: unable to determine the dtExp");

            int strike = Types.RandomDataGenerator.Strike();
            for (int count = 0; count < numPoints; count++, strike += 5)
            {
                //string optionTicker = string.Format("{0} US {1:MM/dd/yy} {2}{3}", ticker, dtExp, (char)optionality, strike);

                Element elm = new ElementReferenceArrayChainTickersItem(ticker, dtExp, optionality, strike);

                base._values.Add(elm);
            }
        }
    }
}

File: BEmu_cpp/src/ReferenceDataRequest/ElementReferenceArrayChainTickers.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/ReferenceDataRequest/ElementReferenceArrayChainTickers.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "ReferenceDataRequest/ElementReferenceArrayChainTickers.h"
#include "ReferenceDataRequest/ElementReferenceArrayChainTickersItem.h"
#include <vector>
#include "Types/RandomDataGenerator.h"
#include "BloombergTypes/Datetime.h"

namespace BEmu
{
	namespace ReferenceDataRequest
	{
		ElementReferenceArrayChainTickers::ElementReferenceArrayChainTickers(const std::string& underlier, unsigned numPoints, const std::string& strDtExp, OptionalityEnum::EnumType optionality) :
			ElementReferenceArray("CHAIN_TICKERS", std::vector<ElementPtr*>())
		{
			int indexSpace = underlier.find(' ');
			std::string ticker = underlier.substr(0, indexSpace); //assume that the underlier takes the form "ABC US EQUITY"

			Datetime dtExp;
			if(strDtExp.length() < 4)
			{
				dtExp = Datetime::Today();
				dtExp.addMonths(1);
				dtExp.addDays(-dtExp.day());
				dtExp.addDays(20); //assume the 20th of the month
			}
			else if(strDtExp.length() == 8)
			{
				dtExp = Datetime::FromYYYYMMDD(strDtExp);
			}
			else if(strDtExp.length() == 6)
			{
				dtExp = Datetime::FromYYMMDD(strDtExp);
			}
			else if (strDtExp[strDtExp.length() - 1] == 'F')
			{
				std::string strNumMonths = strDtExp.substr(0, strDtExp.length() - 1);
				int numMonths = atoi(strNumMonths.c_str());

				dtExp = Datetime::Today();
				dtExp.addMonths(numMonths);
				dtExp.addDays(-dtExp.day());
				dtExp.addDays(20); //assume the 20th of the month
			}
			else
				throw elementPtrEx;

			int strike = RandomDataGenerator::Strike();
			for (int count = 0; count < numPoints; count++, strike += 5)
			{
				ElementPtr * elm = new ElementReferenceArrayChainTickersItem(ticker, dtExp, optionality, strike);
				this->_values.push_back(elm);
			}

		}

	}
}